### PR TITLE
[Not for merge][Feedback Welcome]All-in-one image modifications for ParlAI

### PR DESCRIPTION
**Pull request by dexterju**
**Patch description**
This PR is for collecting feedbacks on how I can get these All-in-One Image-Grounded Conversational Agents https://arxiv.org/abs/1912.12394 paper code merge back to ParlAI. A joint work with @klshuster @ylannb @jaseweston 
This includes several modifications:
1. parlai/core/torch_classifier_agent.py: [A nice plus]
I added codes to load a list of classes file from disk and use them as classification classes on top of the original binary classifier setting.
2. parlai/core/torch_ranker_agent.py [Necessary]
My changes let torch ranker agent take a binary matrix of size [btz * target_classes], this allows it handle multi-label classification. (In the VQA dataset setting, more than one label is considered as correct), so we don't penalize the rest of the correct labels.  
3.  parlai/core/worlds.py[Necessary]
My changes create a new world which flips the multi and batch settings as we had in master. 
4. parlai/utils/misc.py[For compatibility]
For a bug I observed when I tried to load the old model file back. Some previous versions of @klshuster's work packed the opt file with the model file together, and it failed without this line of code.

### CODE REVIEW

File: parlai/core/torch_classifier_agent.py [hunk 1]
```diff
@@ -98,16 +99,18 @@ def __init__(self, opt: Opt, shared=None):
...
-            raise RuntimeError(
-                'Must specify --classes or --classes-from-file argument.'
-            )
         if not shared:
-            if opt['classes_from_file'] is not None:
-                with open(opt['classes_from_file']) as f:
-                    self.class_list = f.read().splitlines()
+            if opt.get('classes') is None:
+                raise RuntimeError('Must specify --classes argument.')
+            if os.path.isfile(opt['classes'][0]):
```
Comment (klshuster) [inline parlai/core/torch_classifier_agent.py]: was this not supported before with the `opt['classes_from_file']`?

File: parlai/core/torch_classifier_agent.py [hunk 2]
```diff
@@ -158,25 +161,26 @@ def __init__(self, opt: Opt, shared=None):
...
+                    )
+                self.model_data_parallel = torch.nn.DataParallel(self.model)
+                self.model = self.model_data_parallel.module
         if shared:
             # We don't use get here because hasattr is used on optimizer later.
             if 'optimizer' in shared:
                 self.optimizer = shared['optimizer']
         else:
-            optim_params = [p for p in self.model.parameters() if p.requires_grad]
+            optim_params = [p for p in self.model.parameters() if p.modus_grad]
```
Comment (klshuster) [inline parlai/core/torch_classifier_agent.py]: what is `modus_grad`?

File: parlai/core/torch_classifier_agent.py [hunk 3]
```diff
@@ -158,25 +161,26 @@ def __init__(self, opt: Opt, shared=None):
...
-                        )
-                    self.model = torch.nn.DataParallel(self.model)
+        if self.use_cuda:
+            if self.opt['data_parallel']:
+                if is_distributed():
+                    raise ValueError(
+                        'Cannot combine --data-parallel and distributed mode'
+                    )
+                self.model_data_parallel = torch.nn.DataParallel(self.model)
+                self.model = self.model_data_parallel.module
```
Comment (klshuster) [inline parlai/core/torch_classifier_agent.py]: does this not mess up later computation? I think we should be calling `self.model()` rather than `self.model_data_parallel()`

File: parlai/core/torch_classifier_agent.py [hunk 4]
```diff
@@ -247,6 +251,7 @@ def train_step(self, batch):
         labels = self._get_labels(batch)
         scores = self.score(batch)
         loss = self.criterion(scores, labels)
+        loss = loss * scores.size(1)
```
Comment (klshuster) [inline parlai/core/torch_classifier_agent.py]: what does this do?

File: parlai/core/torch_classifier_agent.py [hunk 5]
```diff
@@ -269,7 +274,7 @@ def eval_step(self, batch):
             return
 
         self.model.eval()
-        scores = self.score(batch)
+        scores = self.score(batch).float()
```
Comment (klshuster) [inline parlai/core/torch_classifier_agent.py]: can you explain what this fixes too?

File: parlai/core/torch_ranker_agent.py [hunk 6]
```diff
@@ -311,7 +311,7 @@ def _get_train_preds(self, scores, label_inds, cands, cand_vecs):
         else:
             _, ranks = scores.sort(1, descending=True)
         for b in range(batchsize):
-            rank = (ranks[b] == label_inds[b]).nonzero()
+            rank = (ranks[b] == label_inds[b].nonzero()[0]).nonzero()
```
Comment (klshuster) [inline parlai/core/torch_ranker_agent.py]: are these the changes for multi-class classification? does this still work with non-multiclass classification?

File: parlai/core/torch_ranker_agent.py [hunk 7]
```diff
@@ -664,7 +663,9 @@ def _build_candidates(self, batch, source, mode):
                 use_cuda=self.use_cuda,
                 fp16friendly=self.fp16,
             )
-            if label_vecs is not None:
+            if label_vecs is not None and not batch.id.startswith(
+                'internal:vqa_v2:PersonalityTeacher'
```
Comment (klshuster) [inline parlai/core/torch_ranker_agent.py]: i'm not sure we should include this in the actual `torch_ranker_agent` code, perhaps this could be in the model?

File: parlai/core/torch_ranker_agent.py [hunk 8]
```diff
@@ -702,7 +703,9 @@ def _build_candidates(self, batch, source, mode):
             cands = self.fixed_candidates
             cand_vecs = self.fixed_candidate_vecs
 
-            if label_vecs is not None:
+            if label_vecs is not None and not batch.id.startswith(
+                'internal:vqa_v2:PersonalityTeacher'
```
Comment (klshuster) [inline parlai/core/torch_ranker_agent.py]: ditto?

File: parlai/core/worlds.py [hunk 9]
```diff
@@ -656,7 +656,7 @@ def parley_init(self):
         if self.new_world:
             self.new_world = False
             self.parleys = 0
-            if self.random:
+            if False:
```
Comment (klshuster) [inline parlai/core/worlds.py]: what is this change for?

File: parlai/core/worlds.py [hunk 10]
```diff
@@ -839,6 +867,33 @@ def batch_act(self, agent_idx, batch_observation):
...
+    def parley_init(self):
+        """
+        Update the current subworld.
+
+        If we are in the middle of an episode, keep the same world and finish this
+        episode. If we have finished this episode, pick a new world (either in a
+        random or round-robin fashion).
+        """
+        self.parleys = self.parleys + 1
+        if self.world_idx >= 0 and self.worlds[self.world_idx].episode_done():
```
Comment (klshuster) [inline parlai/core/worlds.py]: when is `world_idx` not >= 0?

File: parlai/utils/misc.py [hunk 11]
```diff
@@ -13,6 +13,7 @@
 import time
 from typing import Union, Optional, Set, Any, Dict, List
 import warnings
+from parlai.core.opt import Opt
```
Comment (klshuster) [inline parlai/utils/misc.py]: ahh yes I am aware this broke previous models, we should not add it back here however, those models should be re-saved if necessary (I have already done this with my old models)

File: parlai/core/torch_classifier_agent.py (hunk 1 above)
Comment (emilydinan) [reply parlai/core/torch_classifier_agent.py]: this was indeed supported with the 'classes_from_file' argument

Comment (emilydinan) [reply parlai/core/torch_classifier_agent.py]: i suppose we could overload that argument, i'm not sure how i feel about it. if we decide to do that, then we should get rid of that argument

File: parlai/core/torch_classifier_agent.py (hunk 3 above)
Comment (emilydinan) [reply parlai/core/torch_classifier_agent.py]: yeah i agree. what is going on here? i am pretty sure it worked fine before

File: parlai/core/torch_ranker_agent.py (hunk 7 above)
Comment (emilydinan) [reply parlai/core/torch_ranker_agent.py]: I agree, I do'nt think this should be in TRA

File: parlai/core/worlds.py [hunk 12]
```diff
@@ -747,6 +747,34 @@ def _override_opts_in_shared(table, overrides):
     return table
 
 
+class BatchMultiwWorld(MultiWorld):
```
Comment (emilydinan) [inline parlai/core/worlds.py]: What is this for? can you add a docstring?

File: parlai/core/worlds.py [hunk 13]
```diff
@@ -747,6 +747,34 @@ def _override_opts_in_shared(table, overrides):
...
+            if len(weights) > i:
+                weight = weights[i]
+            else:
+                weight = 1
+            self.cum_task_weights[i] = weight + sum
+            sum += weight
+
+    def parley_init(self):
+        """
+        Flipping world id
```
Comment (emilydinan) [inline parlai/core/worlds.py]: Can you add a more descriptive docstring here? What does "flipping world id" mean?

File: parlai/utils/misc.py (hunk 11 above)
Comment (stephenroller) [reply parlai/utils/misc.py]: Yes, this is resolved in master already

File: parlai/core/torch_classifier_agent.py [hunk 14]
```diff
@@ -346,6 +351,7 @@ def report(self):
         examples = self.metrics['examples']
         if examples > 0:
             m['examples'] = examples
+            m['loss'] = self.metrics['loss']
```
Comment (EricMichaelSmith) [inline parlai/core/torch_classifier_agent.py]: Oh this wasn't used before?

**Comment by dexterju** [comment]
Close, since the issues have been somehow solved.
